### PR TITLE
[ogre-next] Add new port

### PR DESCRIPTION
**Pull request by zi-m**
https://github.com/OGRECave/ogre-next
`# This portfile is based (shamelessly copied and adapted a bit) on 'ogre' portfile.`

### CODE REVIEW

File: ports/ogre-next/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,141 @@
+# This portfile is based (shamelessly copied and adapted a bit) on 'ogre' portfile.
+
+include(vcpkg_common_functions)
+
+if(VCPKG_CMAKE_SYSTEM_NAME AND NOT VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsStore")
```
Comment (vicroms) [inline ports/ogre-next/portfile.cmake]:
```suggestion
if(NOT VCPKG_TARGET_IS_WINDOWS)
```

File: ports/ogre-next/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO OGRECave/ogre-next
+    REF 8083067c1835147de5d82015347d95c710e36bc0
+    SHA512 2660f739def7076b42009ba932a27532f2a20e7be8ec2674aff29116c7dcd7f2d75564d8ce07bef3dba13ece5293d604573ad4f8734f6086c1750127f52834cd
+    HEAD_REF master
+    PATCHES
+        toolchain_fixes.patch
+		rename_config_cmake.patch
+		fix_find_package_sdl2.patch
```
Comment (vicroms) [inline ports/ogre-next/portfile.cmake]: Use only spaces for indentation to keep consistency.

File: ports/ogre-next/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        -DOGRE_BUILD_RENDERSYSTEM_GLES=OFF
+        -DOGRE_BUILD_RENDERSYSTEM_GLES2=OFF
+# Optional stuff
+        -DOGRE_BUILD_COMPONENT_JAVA=${WITH_JAVA}
+        -DOGRE_BUILD_COMPONENT_PYTHON=${WITH_PYTHON}
+        -DOGRE_BUILD_COMPONENT_CSHARP=${WITH_CSHARP}
+        -DOGRE_BUILD_RENDERSYSTEM_D3D9=${WITH_D3D9}
+# vcpkg specific stuff
+        -DOGRE_CMAKE_DIR=share/ogre-next
+)
```
Comment (vicroms) [inline ports/ogre-next/portfile.cmake]: Use `vcpkg_check_features` to simplify the portfile a little bit

```cmake
vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
    d3d9    WITH_D3D9
    java    WITH_JAVA
    python  WITH_PYTHON
    csharp  WITH_CSHARP
)

vcpkg_configure_cmake(
    SOURCE_PATH ${SOURCE_PATH}
    PREFER_NINJA
    OPTIONS 
        -DOGRE_BUILD_DEPENDENCIES=OFF 
        -DOGRE_BUILD_SAMPLES=OFF
        -DOGRE_BUILD_TESTS=OFF 
        -DOGRE_BUILD_TOOLS=OFF
        -DOGRE_BUILD_MSVC_MP=ON
        -DOGRE_BUILD_MSVC_ZM=ON
        -DOGRE_INSTALL_DEPENDENCIES=OFF
        -DOGRE_INSTALL_DOCS=OFF
        -DOGRE_INSTALL_PDB=OFF
        -DOGRE_INSTALL_SAMPLES=OFF
        -DOGRE_INSTALL_TOOLS=OFF
        -DOGRE_INSTALL_CMAKE=ON
        -DOGRE_INSTALL_VSPROPS=OFF
        -DOGRE_STATIC=${OGRE_STATIC}
        -DOGRE_CONFIG_THREAD_PROVIDER=std
        -DOGRE_BUILD_RENDERSYSTEM_D3D11=ON 
        -DOGRE_BUILD_RENDERSYSTEM_GL=ON
        -DOGRE_BUILD_RENDERSYSTEM_GL3PLUS=ON
        -DOGRE_BUILD_RENDERSYSTEM_GLES=OFF
        -DOGRE_BUILD_RENDERSYSTEM_GLES2=OFF
# Optional stuff
       ${FEATURE_OPTIONS}
# vcpkg specific stuff
        -DOGRE_CMAKE_DIR=share/ogre-next
)
```

**Review by vicroms — CHANGES_REQUESTED** [review]

**Comment by traversaro** [comment]
Would it make sense to call the port `ogre2`, that seems to be more informative of the vague `ogre-next` ? This is the same solution used in other repos such as Arch https://aur-dev.archlinux.org/packages/ogre-2.0/ or custom homebrew in https://github.com/osrf/homebrew-simulation/blob/master/Aliases/ogre2 . 

By the way, a custom ogre2 port is available at https://github.com/osrf/vcpkg-ports/tree/master/ogre2 , I do not know if it can be useful for you as reference. 

cc @j-rivero

**Comment by j-rivero** [comment]
> Would it make sense to call the port `ogre2`, that seems to be more informative of the vague `ogre-next` ?

+1 to call it ogre2. The Ogre projects [published some weeks ago the official split](https://www.ogre3d.org/2019/09/15/ogre-ecosystem-roundup-3) between the two projects.

> By the way, a custom ogre2 port is available at https://github.com/osrf/vcpkg-ports/tree/master/ogre2 , I do not know if it can be useful for you as reference.

Our port probably lacks of many features needed for going official, take it with caution. The main goal for us was to get an ogre2 port that was able to be installed side-by-side with ogre1 (we maintain different libraries that has support for both).

**Comment by traversaro** [comment]
> +1 to call it ogre2. The Ogre projects [published some weeks ago the official split](https://www.ogre3d.org/2019/09/15/ogre-ecosystem-roundup-3) between the two projects.

Thanks for the link, I just noticed that there they mention that:

> This also makes it possible to follow semver with Ogre1 and make a 2.0 release when needed.

If there is the possibility in the future of having ogre v2.0 and ogre2 inside vcpkg, that would be quite confusing, and it is something that would suggest that `ogre-next` is a better name. By the way in https://www.ogre3d.org/2019/09/15/ogre-ecosystem-roundup-3 they call the new version `ogre-next`, and they call it `ogre-next` also in https://www.ogre3d.org/about/what-version-to-choose . So perhaps `ogre-next` is indeed the best name for the port.

**Comment by j-rivero** [comment]
> So perhaps `ogre-next` is indeed the best name for the port.

Sounds reasonable to me. I don't have anything strong against it.

File: ports/ogre-next/portfile.cmake (hunk 2 above)
Comment (zi-m) [reply ports/ogre-next/portfile.cmake]: Done.

File: ports/ogre-next/portfile.cmake (hunk 3 above)
Comment (zi-m) [reply ports/ogre-next/portfile.cmake]: Done.

**Comment by PhoebeHui** [comment]
@zi-m, thanks for the PR!
The CI system doesn't test ogre-next features, I tested them manually,  csharp, java and python build successfully . however, d3d9 failed. could you help have a look?

./vcpkg install ogre-next[d3d9]:x64-windows
Failed in cmake configure step. Failures:
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
DirectX9_D3DX9_LIBRARY (ADVANCED)
    linked by target "RenderSystem_Direct3D9" in directory F:/VCPKG/8677/vcpkg/buildtrees/ogre-next/src/c710e36bc0-9cfa542442/RenderSystems/Direct3D9

**Comment by zi-m** [comment]
> @zi-m, thanks for the PR!
> The CI system doesn't test ogre-next features, I tested them manually, csharp, java and python build successfully. However, d3d9 failed. could you help have a look?

@PhoebeHui 
I added environment-overrides.cmake and instructions:
```
Building steps for ogre-next[d3d9]:

1. Download and install "Microsoft DirectX SDK"
https://www.microsoft.com/en-us/download/confirmation.aspx?id=6812

2. Set env variable
set DXSDK_DIR=C:/Program Files (x86)/Microsoft DirectX SDK (June 2010)

3. Install port
.\vcpkg.exe install ogre-next[d3d9] --triplet x64-windows
```

**Review by vicroms — APPROVED** [review]

**Comment by vicroms** [comment]
Thanks for the PR!

**Pull request merged**